changes/ce/feat-12676.en.md

```diff
@@ -0,0 +1,9 @@
+Added new shared subscription strategies
+
+- `sticky_clientid`
+
+    Dispatch messages from a publisher (distinguished by their client ID) always to to the same subscriber as long as both the publisher and subscriber sessions exist. The initial subscriber is selected using the hash_clientid strategy. The difference to hash_clientid is that the subscriber associated to a certain publisher will not change even if new subscribers join the group.
```

Comment:
If I understand the intention of this correctly, you want to make it possible for publisher client to pair up a subscriber client and survive the publisher client reconnect ?

The challenges are
1. The publishing client reconnect will trigger session terminate, and I noticed that session terminate un-sticks the client.
2. When clustered, if a client reconnects, it may connect to another node in the cluster, however the stickiness state is local (in a ETS table).

Reply:
> The main goal is to make sure the subscriber does not change when new subscribers get added to the shared subscription group (or when subscribers are removed, unless it's the current subscriber)

the current sticky strategy should achieve this? Or is it the initial random pick you don’t like?

> I couldn't figure out a better way to make sure that the sticky session data gets cleaned up when it's no longer needed. If the client sets the session expiry interval to a non-zero value the sticky session should stay until that expires regardless of reconnection, right?

process dictionary should do, no need for ets

> Do you think it would make sense to share the sticky state between cluster nodes? What would be the best way to accomplish this?

no, that would be too expensive

Reply:
> The current hash_clientid strategy works well if the subscribers for the group stay constant, but not when they get added/removed (due to kubernetes autoscaling, for example). The hash_clientid strategy picks the Nth subscriber in the list for each message based on the hash but it's not necessarily the same subscriber if the list/count has changed.

I meant the current `sticky` strategy, but did NOT mean the current `hash_clientid` strategy.

> The initial pick is a separate thing, I think it would be nice to have an option for using the subscriber with the least amount of load (which is the goal with the sticky_clientid_leastpubs strategy).

To this, I agree. Maybe worth a separate PR.

> But regardless of which one is used, the sticky state for each client should be cleared when it's no longer needed, right?

Process dictionary goes away when the client process terminates, so there is no need for any cleanup.

Reply:
Hmm. there is perhaps some misunderstanding about how EMQX message dispatch works.
The `sticky` strategy ties to bind a PUBLISH client with a SUBSCRIBE client as long as they are both alive.
If another PUBLISH client sends a message, the `sticky` strategy allows a fresh initialization (random) of the bound.

Same for the process dictionary, the stickiness state is stored in the publisher's process dictionary, but not subscriber.

Reply:
Maybe add a config `initial_sticky_pick = random | hash_clientid | hash_topic | least_loaded`

Reply:
Thank you. I'll have a look soon.